torch/testing/_internal/custom_op_db.py

```diff
@@ -83,9 +86,116 @@ def numpy_take_impl(x, ind, ind_inv, dim):
-    return x
+@numpy_take.impl_fake()
+def numpy_take_fake(x, ind, ind_inv, dim):
+    return torch.empty_like(x)
+
+@custom_op('(Tensor x) -> Tensor', ns='_torch_testing')
+def numpy_nonzero(x):
+    ...
+
+@numpy_nonzero.impl(['cpu', 'cuda'])
+def numpy_nonzero_impl(x):
+    x_np = to_numpy(x)
+    res = np.stack(np.nonzero(x_np), axis=1)
+    if res.shape[0] <= 1:
+        raise RuntimeError("not supported")
+    return torch.tensor(res, device=x.device)
+
+@numpy_nonzero.impl_fake()
+def numpy_nonzero_fake(x):
+    ctx = torch._custom_op.get_ctx()
+    i0 = ctx.new_data_dependent_symint()
+    ctx.constrain_range(i0, min=2)
+    shape = [x.dim(), i0]
+    result = x.new_empty(shape, dtype=torch.long)
+    return result
+
+def sample_inputs_numpy_nonzero(opinfo, device, dtype, requires_grad, **kwargs):
+    make_arg = functools.partial(make_tensor, device=device, dtype=dtype, requires_grad=requires_grad)
+    shape = 10
+    result = make_arg(shape, low=0.9, high=2)
+    mask = make_tensor(shape, low=0, high=2, device=device, dtype=torch.long)
+    with torch.no_grad():
+        result *= mask
+
+    yield SampleInput(result, args=())
+
+
+@custom_op('(Tensor boxes, Tensor scores, float iou_threshold) -> Tensor', ns='_torch_testing')
+def numpy_nms(boxes, scores, iou_threshold):
+    ...
+
+@numpy_nms.impl(['cpu', 'cuda'])
+def numpy_nms_impl(boxes, scores, iou_threshold):
+    # Adapted from Ross Girshick's fast-rcnn implementation at
+    # https://github.com/rbgirshick/fast-rcnn/blob/master/lib/utils/nms.py
+    assert boxes.device == scores.device
+    device = boxes.device
+
+    boxes = to_numpy(boxes)
+    scores = to_numpy(scores)
+
+    N = boxes.shape[0]
+    assert boxes.shape == (N, 4)
+    assert scores.shape == (N,)
+
+    x1 = boxes[:, 0]
+    y1 = boxes[:, 1]
+    x2 = boxes[:, 2]
+    y2 = boxes[:, 3]
+
+    areas = (x2 - x1 + 1) * (y2 - y1 + 1)
+    order = scores.argsort()[::-1]
+
+    keep = []
+    while order.size > 0:
+        i = order[0]
+        keep.append(i)
+        xx1 = np.maximum(x1[i], x1[order[1:]])
+        yy1 = np.maximum(y1[i], y1[order[1:]])
+        xx2 = np.minimum(x2[i], x2[order[1:]])
+        yy2 = np.minimum(y2[i], y2[order[1:]])
+
+        w = np.maximum(0.0, xx2 - xx1 + 1)
+        h = np.maximum(0.0, yy2 - yy1 + 1)
+        inter = w * h
+        ovr = inter / (areas[i] + areas[order[1:]] - inter)
+
+        inds = np.where(ovr <= iou_threshold)[0]
+        order = order[inds + 1]
+
+    result = np.stack(keep)
+    result = torch.tensor(np.stack(keep), device=device)
+    # Needed for data-dependent condition :(
+    assert result.size(0) >= 2
+    return result
+
+@numpy_nms.impl_fake()
+def numpy_nms_fake(boxes, scores, iou_threshold):
+    assert boxes.device == scores.device
+    N = boxes.shape[0]
+    assert boxes.shape == (N, 4)
+    assert scores.shape == (N,)
+
+    ctx = torch._custom_op.get_ctx()
+    i0 = ctx.new_data_dependent_symint()
+    ctx.constrain_range(i0, min=2)
```

Comment:
IMO the API should do this automatically

Reply:
Sounds good, there will just be a warning in the docs that symints must have minimum 2 otherwise things will go wrong.

We still probably want to provide ctx.constrain_range, for those people who have more of an idea what their dynamic dimension produces (e.g. an upper bound for nonzero is the total number of elements in the Tensor)

Reply:
Hmm, I take that back - we only want to constrain the range on symints that get used as the sizes of a Tensor. A user might not take `ctx.new_data_dependent_symint()` and directly pass it as a size to a new Tensor (though this is what most people who will use the API will do).

As an example, someone might have an op `foo(x, y)` that returns a tensor of shape [x_nnz + y_nnz]. One way to write the fake impl is:
```py
def foo_fake(x, y):
    ctx = get_ctx()
    x_nnz = ctx.new_data_dependent_symint()
    y_nnz = ctx.new_data_dependent_symint()
    out_nnz = ctx.constrain_range(x_nnz + y_nnz, 2)
    return x.new_empty([out_nnz])
```
This doesn't really matter right now because we don't do any memoization, but if we were to support memoization, then the user would actually have a reason to create two different symints (so they can be memoized separately)
```py
def foo_fake(x, y):
    ctx = get_ctx()
    x_nnz = ctx.new_data_dependent_symint(memoize_on=(foo_fake, x))
    y_nnz = ctx.new_data_dependent_symint(memoize_on=(foo_fake, y))
    out_nnz = ctx.constrain_range(x_nnz + y_nnz, 2)
    return x.new_empty([out_nnz])
```

Reply:
Resolved offline: we don't care enough, so going to constrain this to 2 by default.